// @kolkov/ngx-dadata/esm5/lib/ngx-dadata.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export var DadataType;
(function (DadataType) {
    DadataType["fio"] = "fio";
    DadataType["address"] = "address";
    DadataType["party"] = "party";
    DadataType["bank"] = "bank";
    DadataType["email"] = "email";
})(DadataType || (DadataType = {}));
var NgxDadataService = /** @class */ (function () {
    function NgxDadataService(http) {
        this.http = http;
        this.apiKey = '';
    }
    NgxDadataService.prototype.setApiKey = function (key) {
        this.apiKey = key;
    };
    // tslint:disable-next-line:max-line-length
    NgxDadataService.prototype.getData = function (value, type, config) {
        if (type === void 0) { type = DadataType.address; }
        var _a, _b, _c, _d, _e, _f, _g;
        var httpOptions = {
            headers: new HttpHeaders({
                Accept: 'application/json',
                'Content-Type': 'application/json',
                Authorization: 'Token ' + this.apiKey,
            })
        };
        var body = Object.assign({}, { query: value }, { count: (_a = config) === null || _a === void 0 ? void 0 : _a.limit }, { location: (_b = config) === null || _b === void 0 ? void 0 : _b.locations }, { location_bust: (_c = config) === null || _c === void 0 ? void 0 : _c.locationsBoost }, { from_bound: (_e = (_d = config) === null || _d === void 0 ? void 0 : _d.bounds) === null || _e === void 0 ? void 0 : _e.fromBound }, { to_bound: (_g = (_f = config) === null || _f === void 0 ? void 0 : _f.bounds) === null || _g === void 0 ? void 0 : _g.toBound });
        return this.http.post('https://suggestions.dadata.ru/suggestions/api/4_1/rs/suggest/' + type, body, httpOptions);
    };
    NgxDadataService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    NgxDadataService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgxDadataService_Factory() { return new NgxDadataService(i0.ɵɵinject(i1.HttpClient)); }, token: NgxDadataService, providedIn: "root" });
    NgxDadataService = __decorate([
        Injectable({
            providedIn: 'root'
        })
    ], NgxDadataService);
    return NgxDadataService;
}());
export { NgxDadataService };
